File: swift/Solutions/Day05.swift
//
//  Day05.swift
//  advent_of_code_2023
//
//  Created by Kushantha Attanayake on 12/3/23.
//

import Foundation

struct Day05: Day {
    func part1Solution(input: [String]) -> String {
        var data = input[0].split(separator: ":")[1].split(separator: " ").map { Int($0)! }
        var rubrick = [Range<Int>:Range<Int>]()
        
        func doDataUpdate(){
            data = data.sorted()
            let sortedRubrick = rubrick.keys.sorted(by: {first, second in first.lowerBound < second.lowerBound})
            var seedIndex = 0
            var rubrickIndex = 0
            while(seedIndex < data.count && rubrickIndex < sortedRubrick.count){
                let curRubrick = sortedRubrick[rubrickIndex]
                let curRubrickMapping = rubrick[curRubrick]!
                let curSeedData = data[seedIndex]
                if(curSeedData > curRubrick.upperBound){
                    rubrickIndex += 1
                } else {
                    // This seed falls within this range
                    if(curSeedData >= curRubrick.lowerBound && curSeedData <= curRubrick.upperBound){
                        data[seedIndex] = curRubrickMapping.lowerBound + (curSeedData - curRubrick.lowerBound)
                    }
                    seedIndex += 1
                }
            }
        }
        
        input[2...].forEach { line in
            if(line.contains("map:")){ // Done parsing rubrick, can map data
                doDataUpdate()
                rubrick.removeAll()
            } else {
                let rubrickParts = line.split(separator: " ")
                let destStart = Int(rubrickParts[0])!
                let sourceStart = Int(rubrickParts[1])!
                let rangeLength = Int(rubrickParts[2])!
                rubrick[sourceStart..<(sourceStart+rangeLength)] = destStart..<(destStart+rangeLength)
            }
        }
        doDataUpdate()
        return "\(data.min()!)"
    }
    
    func part2Solution(input: [String]) -> String {
        let dataParts = input[0].split(separator: ":")[1].split(separator: " ").map { Int($0)! }
        var data = (0..<(dataParts.count/2)).map {
            let startIndex = $0 * 2
            let endIndex = startIndex + 1
            return dataParts[startIndex]..<(dataParts[startIndex]+dataParts[endIndex])
        }
        var rubrick = [Range<Int>:Range<Int>]()
        
        func doDataUpdate(){
            data = data.sorted(by: {first, second in first.lowerBound < second.lowerBound})
            let sortedRubrick = rubrick.keys.sorted(by: {first, second in first.lowerBound < second.lowerBound})
            var seedIndex = 0
            var rubrickIndex = 0
            while(seedIndex < data.count && rubrickIndex < sortedRubrick.count){
                let curRubrick = sortedRubrick[rubrickIndex]
                let curRubrickMapping = rubrick[curRubrick]!
                let curSeedData = data[seedIndex]
                if(curSeedData.lowerBound > (curRubrick.upperBound-1)){ // no overlap AND rubrick < seed range
                    rubrickIndex += 1
                } else { // Rubrik not completely smaller than seed range
                    
                    if(curSeedData.lowerBound >= curRubrick.lowerBound  && curSeedData.upperBound < curRubrick.upperBound){
                        // Full overlap scenario
                        let newMappingStart = curRubrickMapping.lowerBound + (curSeedData.lowerBound - curRubrick.lowerBound)
                        let newMappingEnd = newMappingStart + curSeedData.count
                        data[seedIndex] = newMappingStart..<newMappingEnd
                        
                    } else if (curSeedData.lowerBound >= curRubrick.lowerBound && curSeedData.lowerBound < curRubrick.upperBound){
                        // Left overlap
                        let newMappingStart = curRubrickMapping.lowerBound + (curSeedData.lowerBound - curRubrick.lowerBound)
                        let newMappingEnd = curRubrickMapping.upperBound
                        data[seedIndex] = newMappingStart..<newMappingEnd
                        
                        let leftoverStart = curRubrick.upperBound
                        let leftoverEnd = curSeedData.upperBound
                        data.insert(leftoverStart..<leftoverEnd, at: seedIndex+1)
                        
                    } else if (curSeedData.upperBound < curRubrick.upperBound && curSeedData.upperBound > curRubrick.lowerBound) {
                        let newMappingStart = curRubrickMapping.lowerBound
                        let newMappingEnd = newMappingStart + (curSeedData.upperBound - curRubrick.lowerBound)
                        data[seedIndex] = newMappingStart..<newMappingEnd
                        
                        let leftoverStart = curSeedData.lowerBound
                        let leftoverEnd = curRubrick.lowerBound
                        data.insert(leftoverStart..<leftoverEnd, at: seedIndex)
                        seedIndex += 1
                    }
                    // Partial overlap on left
                    // Partial overlap on right
                    // This seed falls within this range
                    seedIndex += 1
                }
            }
        }
        
        input[2...].forEach { line in
            if(line.contains("map:")){ // Done parsing rubrick, can map data
                doDataUpdate()
                rubrick.removeAll()
            } else {
                let rubrickParts = line.split(separator: " ")
                let destStart = Int(rubrickParts[0])!
                let sourceStart = Int(rubrickParts[1])!
                let rangeLength = Int(rubrickParts[2])!
                rubrick[sourceStart..<(sourceStart+rangeLength)] = destStart..<(destStart+rangeLength)
            }
        }
        doDataUpdate()
        return "\(data.sorted(by: {first, second in first.lowerBound < second.lowerBound})[0].lowerBound)"
    }
    
    
}

///
///25834266
///97676188
///486613012
///
///

File: kotlin/src/main/kotlin/solutions/Day05.kt
package solutions

class Day05: Day {
    override fun part1Solution(input: List<String>): String {
        val data = input[0].split(":")[1].trim().split(" ").map { it.toLong() }.toMutableList()
        val rubric = mutableMapOf<LongRange, LongRange>()

        fun doDataUpdate(){
            data.sort()
            val sortedRubric = rubric.keys.sortedBy { it.first }
            var seedIndex = 0
            var rubricIndex = 0
            while(seedIndex < data.size && rubricIndex < sortedRubric.size){
                val curRubric = sortedRubric[rubricIndex]
                val curRubricMapping = rubric[curRubric]!!
                val curSeedData = data[seedIndex]
                if(curSeedData > curRubric.last) rubricIndex++
                else {
                    // This seed falls within this range
                    if(curSeedData >= curRubric.first){
                        data[seedIndex] = curRubricMapping.first + (curSeedData - curRubric.first)
                    }
                    seedIndex++
                }
            }
        }

        input.slice(3..<input.size).forEach { line ->
                if(line.contains("map:")){ // Done parsing rubric, can map data
                    doDataUpdate()
                    rubric.clear()
                } else if(line.isNotEmpty()) {
                    val rubricParts = line.split( " ")
                    val destStart = rubricParts[0].toLong()
                    val sourceStart = rubricParts[1].toLong()
                    val rangeLength = rubricParts[2].toLong()
                    rubric[sourceStart..<(sourceStart+rangeLength)] = destStart..<(destStart+rangeLength)
                }
        }
        doDataUpdate()
        return data.min().toString()
    }

    override fun part2Solution(input: List<String>): String {
        val dataParts = input[0].split(":")[1].trim().split(" ").map { it.toLong() }
        val data = (0..<(dataParts.size/2)).map {
            val startIndex = it * 2
            val endIndex = startIndex + 1
            dataParts[startIndex]..<(dataParts[startIndex]+dataParts[endIndex])
        }.toMutableList()
        val rubric = mutableMapOf<LongRange, LongRange>()

        fun doDataUpdate(){
            data.sortBy { it.first }
            val sortedRubric = rubric.keys.sortedBy { it.first }
            var seedIndex = 0
            var rubricIndex = 0
            while(seedIndex < data.size && rubricIndex < sortedRubric.size){
                val curRubric = sortedRubric[rubricIndex]
                val curRubricMapping = rubric[curRubric]!!
                val curSeedData = data[seedIndex]
                if(curSeedData.first > (curRubric.last)){ // no overlap AND rubric < seed range
                    rubricIndex++
                } else { // Rubric not completely smaller than seed range

                    if(curSeedData.first >= curRubric.first  && curSeedData.last <= curRubric.last){
                        // Full overlap scenario
                        val newMappingStart = curRubricMapping.first + (curSeedData.first - curRubric.first)
                        val newMappingEnd = newMappingStart + curSeedData.length
                        data[seedIndex] = newMappingStart..<newMappingEnd

                    } else if (curSeedData.first >= curRubric.first && curSeedData.first <= curRubric.last){
                        // Left overlap
                        val newMappingStart = curRubricMapping.first + (curSeedData.first - curRubric.first)
                        val newMappingEnd = curRubricMapping.last + 1
                        data[seedIndex] = newMappingStart..<newMappingEnd

                        val leftoverStart = curRubric.last + 1
                        val leftoverEnd = curSeedData.last + 1
                        data.add(seedIndex+1, leftoverStart..<leftoverEnd)

                    } else if (curSeedData.last <= curRubric.last && curSeedData.last >= curRubric.first) {
                        // Right overlap
                        val newMappingStart = curRubricMapping.first
                        val newMappingEnd = newMappingStart + (curSeedData.last - curRubric.first - 1)
                        data[seedIndex] = newMappingStart..<newMappingEnd

                        val leftoverStart = curSeedData.first
                        val leftoverEnd = curRubric.first
                        data.add(seedIndex, leftoverStart..<leftoverEnd)
                        seedIndex++
                    }
                    seedIndex++
                }
            }
        }

        input.slice(3..<input.size).forEach { line ->
                if(line.contains("map:")){ // Done parsing rubric, can map data
                    doDataUpdate()
                    rubric.clear()
                } else if(line.isNotEmpty()) {
                    val rubricParts = line.split(" ")
                    val destStart = rubricParts[0].toLong()
                    val sourceStart = rubricParts[1].toLong()
                    val rangeLength = rubricParts[2].toLong()
                    rubric[sourceStart..<(sourceStart+rangeLength)] = destStart..<(destStart+rangeLength)
                }
        }
        doDataUpdate()
        return data.minByOrNull { it.first }!!.first.toString()
    }
}

private val LongRange.length
    get() = last - first + 1